compiler/rustc_mir_transform/src/check_unnecessary_transmutes.rs

```diff
@@ -88,7 +91,17 @@ impl<'tcx> Visitor<'tcx> for UnnecessaryTransmuteChecker<'_, 'tcx> {
             && self.tcx.is_intrinsic(func_def_id, sym::transmute)
             && let span = self.body.source_info(location).span
             && let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(arg)
-            && let Some(lint) = self.is_unnecessary_transmute(func, snippet, span)
+            && let def_id = self.body.source.def_id()
+            && let Some(lint) = self.is_unnecessary_transmute(
+                func,
+                snippet,
+                span,
+                self.tcx.is_const_fn(def_id)
```

Comment:
I don't believe this heuristic is complete. This leaves out const items and statics, for example.

Reply:
could you add some tests for statics, consts, and in const blocks?

Reply:
im not sure the heuristic is complete, but i added those tests and changed it a little.